src/lib/HST/Util/Selectors.hs

```diff
@@ -89,3 +93,106 @@ getPatVarName (S.PTuple _ _ _) = reportFatal
+    ]
+
+-- | Collects all identifiers in a pattern.
+instance HasIdentifiers (S.Pat a) where
+  findIdentifiers (S.PVar _ name)                 = findIdentifiers name
+  findIdentifiers (S.PInfixApp _ pat1 qname pat2) = findIdentifiers qname
+    `Set.union` (findIdentifiers pat1 `Set.union` findIdentifiers pat2)
+  findIdentifiers (S.PApp _ qname pats)
+    = findIdentifiers qname `Set.union` findIdentifiers pats
+  findIdentifiers (S.PTuple _ _ pats)             = findIdentifiers pats
+  findIdentifiers (S.PParen _ pat)                = findIdentifiers pat
+  findIdentifiers (S.PList _ pats)                = findIdentifiers pats
+  findIdentifiers (S.PWildCard _)                 = Set.empty
+
+-- | Collects all identifiers in a right-hand side of a pattern matching rule.
+instance HasIdentifiers (S.Rhs a) where
+  findIdentifiers (S.UnGuardedRhs _ expr) = findIdentifiers expr
+  findIdentifiers (S.GuardedRhss _ grhss) = findIdentifiers grhss
+
+-- | Collects all identifiers in a guarded right-hand side.
+instance HasIdentifiers (S.GuardedRhs a) where
+  findIdentifiers (S.GuardedRhs _ cond expr)
+    = findIdentifiers cond `Set.union` findIdentifiers expr
+
+-- | Collects all identifiers in an expression.
+instance HasIdentifiers (S.Exp a) where
+  findIdentifiers (S.Var _ qname)            = findIdentifiers qname
+  findIdentifiers (S.Con _ qname)            = findIdentifiers qname
+  findIdentifiers (S.Lit _ _)                = Set.empty
+  findIdentifiers (S.InfixApp _ exp1 _ exp2)
+    = findIdentifiers exp1 `Set.union` findIdentifiers exp2
+  findIdentifiers (S.App _ exp1 exp2)
+    = findIdentifiers exp1 `Set.union` findIdentifiers exp2
+  findIdentifiers (S.NegApp _ expr)          = findIdentifiers expr
+  findIdentifiers (S.Lambda _ pats expr)
+    = findIdentifiers pats `Set.union` findIdentifiers expr
+  findIdentifiers (S.Let _ binds expr)
+    = findIdentifiers binds `Set.union` findIdentifiers expr
+  findIdentifiers (S.If _ cond exp1 exp2)    = findIdentifiers cond
+    `Set.union` findIdentifiers exp1
+    `Set.union` findIdentifiers exp2
+  findIdentifiers (S.Case _ scrutinee alts)
+    = findIdentifiers scrutinee `Set.union` findIdentifiers alts
+  findIdentifiers (S.Tuple _ _ exps)         = findIdentifiers exps
+  findIdentifiers (S.List _ exps)            = findIdentifiers exps
+  findIdentifiers (S.Paren _ expr)           = findIdentifiers expr
+  findIdentifiers (S.ExpTypeSig _ expr _)    = findIdentifiers expr
+
+-- | Collects all identifiers in a @case@ alternative.
+instance HasIdentifiers (S.Alt a) where
+  findIdentifiers (S.Alt _ pat rhs binds) = findIdentifiers pat
+    `Set.union` findIdentifiers rhs
+    `Set.union` findIdentifiers binds
+
+-- | Collects all identifiers in @let@ and @where@ bindings.
+instance HasIdentifiers (S.Binds a) where
+  findIdentifiers (S.BDecls _ decls) = findIdentifiers decls
+
+instance HasIdentifiers (S.QName a) where
+  findIdentifiers (S.Qual _ _ name) = findIdentifiers name
```

Comment:
I'm not too sure whether we should include identifiers if they are qualified. If an expression contains a name `M.a1` (but not the unqualified name `a1`), there is no name conflict with a fresh variable `a1`. However, it wouldn't hurt if we rename the fresh variable nontheless.

Reply:
I think we should choose different names to make it less confusing for the user.